Ramble reports

dapomeroy opened the pull request.

dapomeroy committed:
Add ramble report command to generate charts

Adds a new command `ramble report` to generate charts from experiment
data.

dapomeroy committed:
Add FOM types

Adds FOM type to figure_of_merit directive

dapomeroy committed:
Import results files for ramble report

Adds ability to import results files by filename or workspace name for
ramble report

dapomeroy committed:
Add config setting for report path

Adds a config setting for report folder path

dapomeroy committed:
Troubleshooting fom_type

dapomeroy committed:
Add test data for reports

Adds sample JSON and YAML results data for use in reports tests

dapomeroy committed:
Add unit tests for Ramble reports

Commit:
allow workspace to be read from json

Commit:
move plots to class and factory

rfbgo committed:
Abstraction normalization and scaling

rfbgo committed:
Add basic unit test and results variance plotting

rfbgo committed:
add logic around repeat children

rfbgo committed:
split draw into seperate function

rfbgo committed:
Add unit tests and remove duplicaiton

rfbgo committed:
Add unit tests and remove duplicaiton

rfbgo committed:
add minmax for stats on weak scaling

rfbgo committed:
add log axis

rfbgo committed:
add example multi line plot

rfbgo committed:
Add where and disallow multi-line scaling plots

rfbgo committed:
unify data generation between strong and weak scaling

rfbgo committed:
unify draw logic

rfbgo committed:
add fom plot base line

rfbgo committed:
remove double numeric cast

rfbgo committed:
Adopt new JSON child format

dapomeroy committed:
Apply black formatting to reports and results

dapomeroy committed:
Fix normalization divide by zero issue

Fixes divide by zero error when normalizing series beginning with zero and
inserts filler chart when plotting fails. Adjusts chart size, legend,
and axis ticks for clarity.

dapomeroy committed:
Create one instance per spec, store series calcs

Updates to generate_plot_data to enable multi-line plots. Removes spec
loop from PlotGenerator instance and creates one instance per spec.
Stores calculations for each series in instance output_df so they can be
used to plot multiple series on the same chart.

dapomeroy committed:
Add multi-line plot creation

Adds multi-line plots. Fixes y-axis for normalization of scaling plots
with the first scaling value >1.

dapomeroy committed:
Add better direction suffix to chart title

dapomeroy committed:
Fix x-axis title falling off chart

Fixes issue with x-axis title formatting

dapomeroy committed:
Misc formatting fixes and cleanup

dapomeroy committed:
DRY minmax calculations

dapomeroy committed:
Fix formatting

dapomeroy committed:
prevent calcuation of repeat statistics on INFO and CATEGORY fom types

rfbgo commented on lib/ramble/ramble/cmd/results.py:
```diff
@@ -36,25 +133,57 @@ def results_upload(args):
     logger.debug("File to import:")
     logger.debug(filename)
 
+    # TODO: this isn't very safe checking
```
Can we wrap this in a `with` or (`try`) block please?

dapomeroy replied:
Done

rfbgo commented on lib/ramble/ramble/language/shared_language.py:
```diff
@@ -77,8 +79,57 @@ def _execute_figure_of_merit_context(obj):
+        if self == BetterDirection.LOWER:
+            return " (Lower is Better)"
+        else:
+            return ""
```
Should this be `INDETERMINATE`?

dapomeroy replied:
This method returns a string that we can append on a chart title to convey the 'better direction' in the chart text. For example, with Gromacs you might get "Core Time by workload name (Lower is Better)" as the chart title. But if better direction is indeterminate I don't think we should append anything on the title, hence the empty string.

dapomeroy replied:
Per discussion, moved this out of the class and into reports.py

rfbgo commented on lib/ramble/ramble/language/shared_language.py:
```diff
@@ -77,8 +79,57 @@ def _execute_figure_of_merit_context(obj):
+
+        return direction[self]
+
+    def copy(self):
```
Whats the motivation behind doing this instead of expecting a user to do something like a deepcopy?

dapomeroy replied:
I was trying to avoid making a special case out of FomType in this loop within ApplicationBase._analysis_dicts() (L2102) so that it can copy the FomType object:
```
            for fom, fom_def in mod.figures_of_merit.items():
                fom_definitions[fom] = {"origin": f"{mod}", "origin_type": "modifier"}
                for attr in fom_def.keys():
                    if isinstance(fom_def[attr], (list, FomType)):
                        fom_definitions[fom][attr] = fom_def[attr].copy()
                    else:
                        fom_definitions[fom][attr] = self.expander.expand_var(
                            fom_def[attr], mod_vars
                        )
```

I saw several examples of this kind of copy method in the code already, like WorkloadVariable. Do you know of a better way to do it?

dapomeroy replied:
Updated to deepcopy to match the other class methods

rfbgo commented on lib/ramble/ramble/reports.py:
```diff
@@ -0,0 +1,959 @@
+
+import pandas as pd
+
+# import numpy as np
```
remove?

rfbgo commented on lib/ramble/ramble/reports.py:
```diff
@@ -0,0 +1,959 @@
+
+def is_repeat_child(experiment):
+    try:
+        if int(experiment["RAMBLE_VARIABLES"]["repeat_index"]) > 0:
```
Can we update this to use the recently added/propogated `repeat_index`? 

(see https://github.com/GoogleCloudPlatform/ramble/pull/633)

dapomeroy replied:
Done

rfbgo commented on lib/ramble/ramble/reports.py:
```diff
@@ -0,0 +1,959 @@
+                    exp_copy = copy.deepcopy(exp)
+
+                    # Remove context dict and add the current FOM values
+                    # TODO: Can we clean this partial copy with a class?
```
Any reason this isn't something more like a copy?

dapomeroy replied:
Let's discuss. I can see some drawbacks here but not quite clear on how to improve it.

dapomeroy replied:
Updated with a mapping dict to remove some of the magic strings.

dapomeroy replied:
Added enums to remove most of the rest of magic strings

rfbgo commented on lib/ramble/ramble/test/data/reports/results.test.json:
```diff
@@ -0,0 +1 @@
+/usr/local/google/home/dpomeroy/VSCode/ramble/var/ramble/workspaces/basic_gromacs/results.2024-06-28_20.55.29.json
```
Not a big deal, but do you want such a "real" (and personal) path here?

dapomeroy replied:
Oh, this is weird. I have a full results file that's been de-identified in my branch, but it's showing up here as just one line. I must have mixed something up in Git. I'll fix it - please disregard for the moment. Thanks!

dapomeroy replied:
Figured it out. Somehow I ended up with a symlink instead of the file inside my workspace. Now fixed

rfbgo commented on pytest.ini:
```diff
@@ -3,7 +3,7 @@
 testpaths =
   lib/ramble/ramble/test
-  var/ramble/repos/*/*/*/test
+  #var/ramble/repos/*/*/*/test/*.py
```
Just double checking, is this intended?

dapomeroy replied:
I don't know. You changed it in this commit: https://github.com/GoogleCloudPlatform/ramble/commit/7f8ab379ff063abf783b04f2d86531e333a53738 
Do you remember what it was about? Can I just assume it was to test something out and revert it?

dapomeroy replied:
Reverted

rfbgo reviewed (changes requested):
Some quick feedback and suggestions, since I know this is not yet in it's final form 

Thanks for working on this awesome feature!

dapomeroy committed:
Address reviewer feedback

dapomeroy committed:
Replace magic strings with enums

douglasjacobsen commented on lib/ramble/ramble/language/shared_language.py:
```diff
@@ -77,8 +81,66 @@ def _execute_figure_of_merit_context(obj):
 
 
+# For a FOM, the direction that is 'better' e.g., faster is better
+class BetterDirection(Enum):
```
This is super nit-picky, but I would kind of prefer if these (BetterDirection and FomType) lived in a separate module that is more clear it relates to FOMs.

Maybe in a util module or something?

IMO, I would almost never think to look in the shared_language module for enums about how FOMs behave.

dapomeroy replied:
OK, moved out to a new foms.py util module

douglasjacobsen commented on lib/ramble/ramble/application.py:
```diff
@@ -1824,11 +1825,17 @@ def calculate_statistics(self, workspace):
 
-        def is_numeric(value):
-            """Returns true if a fom value is numeric"""
+        def is_numeric_fom(fom):
```
Could it make sense to move this into the FomType class? I'm not sure if it makes the most sense yet...

Maybe at a minimum put a TODO to help us remember to move it once we create a Fom class.

dapomeroy replied:
At first glance I don't find it intuitive, because the FomType class is an enum that doesn't have any information about the FOM inside of it. I think a FOM class seems like a better fit. Added todo.

douglasjacobsen commented on lib/ramble/ramble/cmd/results.py:
```diff
@@ -6,25 +6,122 @@
+        "[optional: group by]",
+        required=False,
+    )
+    report_parser.add_argument(
```
This might be answered as I get through more of the code, but if this is used for filtering how some experiments...
You can use the where common argument (in the common module) for this.

Also, you might want to add  the `--execlude-where` argument to keep the interfaces symmetric.

dapomeroy replied:
Bob added this and I think it works differently from the common 'where'. It passes in a 'where' string to pandas to downselect the dataframe. It's more like a SQL 'where' statement. I haven't looked into this closely to see if the common 'where' might be useful, but making the pandas query format available adds a lot of control over what shows up in the chart.

rfbgo replied:
Will rename to `pandas-where`

douglasjacobsen commented on lib/ramble/ramble/cmd/results.py:
```diff
@@ -36,25 +133,56 @@ def results_upload(args):
 
-    imported_file = open(filename)
+    with open(filename) as imported_file:
+        logger.msg(f"Importing {filename}")
```
Maybe update to `Importing results file: {filename}`

dapomeroy replied:
Done

douglasjacobsen commented on requirements.txt:
```diff
@@ -8,3 +8,5 @@ pyarrow==3.0.0;python_version<="3.6"
 google-cloud-bigquery
 tqdm
 deprecation
+pandas
```
We discussed this internally, but each of these should have a guarded import in the code, where if the import fails we print something about installing the requirements.txt file.

dapomeroy replied:
This is done

dapomeroy committed:
Add deps to cloud build trigger

dapomeroy committed:
Elevate non-numeric FOMs into repeat summaries

Adds non-numeric FOMS from repeated experiments into the repeat base,
if identical on all repeats. This fixes an issue where metadata from
modifiers was not available for repeat summary data.

dapomeroy committed:
Add test for repeated exp summary data in reports

Tests that repeat data is correctly imported for reports. Updates
existing report test.

dapomeroy committed:
Address reviewer feedback

Fixes minor issues, moves FomType and BetterDirection to new module
